File: Core/StateDatabase.cs
using System.Text.Json;
using Microsoft.Data.Sqlite;

namespace EpikChatBot.Core;

public sealed class StateDatabase : IDisposable
{
  private readonly SqliteConnection _connection;
  private readonly object _gate = new();

  public StateDatabase(string filePath)
  {
    filePath = Path.GetFullPath(filePath);
    Directory.CreateDirectory(Path.GetDirectoryName(filePath)!);
    ApplyPendingRestore(filePath);
    _connection = new SqliteConnection($"Data Source={filePath}");
    _connection.Open();
    Execute("PRAGMA journal_mode=WAL; PRAGMA synchronous=NORMAL; PRAGMA foreign_keys=ON; PRAGMA busy_timeout=5000;");
    Execute("""
            CREATE TABLE IF NOT EXISTS schema_migrations(version INTEGER PRIMARY KEY, applied_at INTEGER NOT NULL);
            CREATE TABLE IF NOT EXISTS documents(key TEXT PRIMARY KEY, value TEXT NOT NULL, updated_at INTEGER NOT NULL);
            CREATE TABLE IF NOT EXISTS records(namespace TEXT NOT NULL, key TEXT NOT NULL, value TEXT NOT NULL, updated_at INTEGER NOT NULL, PRIMARY KEY(namespace,key));
            INSERT OR IGNORE INTO schema_migrations(version,applied_at) VALUES(1,unixepoch('now')*1000);
            """);
  }

  public static bool ApplyPendingRestore(string path)
  {
    var pending = $"{Path.GetFullPath(path)}.restore-pending";
    if (!File.Exists(pending)) return false;
    Directory.CreateDirectory(Path.GetDirectoryName(Path.GetFullPath(path))!);
    if (File.Exists(path)) File.Copy(path, $"{path}.pre-restore.bak", true);
    File.Move(pending, path, true);
    return true;
  }

  public T Get<T>(string nameSpace, string key, T fallback)
  {
    lock (_gate)
    {
      using var command = _connection.CreateCommand();
      command.CommandText = "SELECT value FROM records WHERE namespace=$namespace AND key=$key";
      command.Parameters.AddWithValue("$namespace", nameSpace); command.Parameters.AddWithValue("$key", key);
      if (command.ExecuteScalar() is not string value) return fallback;
      try { return JsonSerializer.Deserialize<T>(value, JsonStore.Options) ?? fallback; }
      catch (Exception exception) { Console.Error.WriteLine($"[database] Unable to deserialize {nameSpace}/{key}: {exception.Message}"); return fallback; }
    }
  }

  public T Set<T>(string nameSpace, string key, T value)
  {
    lock (_gate)
    {
      using var command = _connection.CreateCommand();
      command.CommandText = "INSERT INTO records(namespace,key,value,updated_at) VALUES($namespace,$key,$value,$at) ON CONFLICT(namespace,key) DO UPDATE SET value=excluded.value,updated_at=excluded.updated_at";
      command.Parameters.AddWithValue("$namespace", nameSpace); command.Parameters.AddWithValue("$key", key);
      command.Parameters.AddWithValue("$value", JsonSerializer.Serialize(value, JsonStore.Options)); command.Parameters.AddWithValue("$at", DateTimeOffset.UtcNow.ToUnixTimeMilliseconds());
      command.ExecuteNonQuery(); return value;
    }
  }

  public bool Delete(string nameSpace, string key)
  {
    lock (_gate)
    {
      using var command = _connection.CreateCommand(); command.CommandText = "DELETE FROM records WHERE namespace=$namespace AND key=$key";
      command.Parameters.AddWithValue("$namespace", nameSpace); command.Parameters.AddWithValue("$key", key); return command.ExecuteNonQuery() > 0;
    }
  }

  public IReadOnlyList<(string Key, T Value, long UpdatedAt)> List<T>(string nameSpace)
  {
    lock (_gate)
    {
      var result = new List<(string, T, long)>(); using var command = _connection.CreateCommand();
      command.CommandText = "SELECT key,value,updated_at FROM records WHERE namespace=$namespace ORDER BY updated_at"; command.Parameters.AddWithValue("$namespace", nameSpace);
      using var reader = command.ExecuteReader(); while (reader.Read()) { try { var value = JsonSerializer.Deserialize<T>(reader.GetString(1), JsonStore.Options); if (value is not null) result.Add((reader.GetString(0), value, reader.GetInt64(2))); } catch (Exception exception) { Console.Error.WriteLine($"[database] Unable to deserialize {nameSpace}/{reader.GetString(0)}: {exception.Message}"); } }
      return result;
    }
  }

  public void Checkpoint() => Execute("PRAGMA wal_checkpoint(TRUNCATE)");
  private void Execute(string sql) { lock (_gate) { using var command = _connection.CreateCommand(); command.CommandText = sql; command.ExecuteNonQuery(); } }
  public void Dispose() { Checkpoint(); _connection.Dispose(); }
}

File: legacy-node/lib/state-database.js
const fs = require("fs");
const path = require("path");
const { DatabaseSync } = require("node:sqlite");
const { applyMigrations } = require("./database-migrations");

class StateDatabase {
  static applyPendingRestore(filePath, logger = console) {
    const resolved = path.resolve(filePath); const pending = `${resolved}.restore-pending`;
    if (!fs.existsSync(pending)) return false;
    fs.mkdirSync(path.dirname(resolved), { recursive: true });
    if (fs.existsSync(resolved)) fs.copyFileSync(resolved, `${resolved}.pre-restore.bak`);
    fs.renameSync(pending, resolved);
    logger.warn("[database] Applied pending restore; previous database saved with .pre-restore.bak");
    return true;
  }

  constructor({ filePath, logger = console }) {
    this.filePath = path.resolve(filePath);
    this.logger = logger;
    fs.mkdirSync(path.dirname(this.filePath), { recursive: true });
    this.db = new DatabaseSync(this.filePath);
    this.db.exec("PRAGMA journal_mode=WAL; PRAGMA synchronous=NORMAL; PRAGMA foreign_keys=ON; PRAGMA busy_timeout=5000;");
    this.schemaVersion = applyMigrations(this.db, this.logger);
    this.readDocumentStatement = this.db.prepare("SELECT value,updated_at FROM documents WHERE key = ?");
    this.writeDocumentStatement = this.db.prepare("INSERT INTO documents(key,value,updated_at) VALUES(?,?,?) ON CONFLICT(key) DO UPDATE SET value=excluded.value, updated_at=excluded.updated_at");
    this.getRecordStatement = this.db.prepare("SELECT value FROM records WHERE namespace = ? AND key = ?");
    this.setRecordStatement = this.db.prepare("INSERT INTO records(namespace,key,value,updated_at) VALUES(?,?,?,?) ON CONFLICT(namespace,key) DO UPDATE SET value=excluded.value, updated_at=excluded.updated_at");
    this.deleteRecordStatement = this.db.prepare("DELETE FROM records WHERE namespace = ? AND key = ?");
    this.listRecordStatement = this.db.prepare("SELECT key,value,updated_at FROM records WHERE namespace = ? ORDER BY updated_at ASC");
  }

  documentKey(filePath) { return path.resolve(filePath).toLowerCase(); }

  readDocument(filePath, fallback) {
    const key = this.documentKey(filePath);
    const row = this.readDocumentStatement.get(key);
    if (row) {
      const fileUpdatedAt = fs.existsSync(filePath) ? fs.statSync(filePath).mtimeMs : 0;
      if (fileUpdatedAt > row.updated_at + 1000) {
        try {
          const restored = JSON.parse(fs.readFileSync(filePath, "utf8"));
          this.writeDocument(filePath, restored);
          this.logger.log("[database] Imported newer restored JSON", path.basename(filePath));
          return restored;
        } catch (error) { this.logger.error("[database] Unable to import restored JSON", { filePath, error: error.message }); }
      }
      try { return JSON.parse(row.value); } catch (error) { this.logger.error("[database] Invalid stored document", { key, error: error.message }); return fallback; }
    }
    if (!fs.existsSync(filePath)) return fallback;
    try {
      const value = JSON.parse(fs.readFileSync(filePath, "utf8"));
      this.writeDocument(filePath, value);
      this.logger.log("[database] Imported legacy JSON", path.basename(filePath));
      return value;
    } catch (error) {
      this.logger.error("[database] Unable to import legacy JSON", { filePath, error: error.message });
      return fallback;
    }
  }

  writeDocument(filePath, value) {
    this.writeDocumentStatement.run(this.documentKey(filePath), JSON.stringify(value), Date.now());
  }

  get(namespace, key, fallback = null) {
    const row = this.getRecordStatement.get(String(namespace), String(key));
    if (!row) return fallback;
    try { return JSON.parse(row.value); } catch { return fallback; }
  }

  set(namespace, key, value) {
    this.setRecordStatement.run(String(namespace), String(key), JSON.stringify(value), Date.now());
    return value;
  }

  delete(namespace, key) { return this.deleteRecordStatement.run(String(namespace), String(key)).changes > 0; }

  list(namespace) {
    return this.listRecordStatement.all(String(namespace)).flatMap((row) => {
      try { return [{ key: row.key, value: JSON.parse(row.value), updatedAt: row.updated_at }]; } catch { return []; }
    });
  }

  transaction(callback) {
    this.db.exec("BEGIN IMMEDIATE");
    try { const result = callback(); this.db.exec("COMMIT"); return result; }
    catch (error) { this.db.exec("ROLLBACK"); throw error; }
  }
  checkpoint() { this.db.exec("PRAGMA wal_checkpoint(TRUNCATE)"); }
  getSchemaVersion() { return this.schemaVersion; }
  close() { this.checkpoint(); this.db.close(); }
}

module.exports = { StateDatabase };
